src/librustc_mir/util/pretty.rs

```diff
@@ -534,6 +538,201 @@ pub fn write_mir_intro<'tcx>(
     Ok(())
 }
 
+/// Find all `AllocId`s mentioned (recursively) in the MIR body and print their corresponding
+/// allocations.
+pub fn write_allocations<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    body: &Body<'_>,
+    w: &mut dyn Write,
+) -> io::Result<()> {
+    fn alloc_ids_from_alloc(alloc: &Allocation) -> impl Iterator<Item = AllocId> + '_ {
+        alloc.relocations().iter().map(|(_, (_, id))| *id)
+    }
+    fn alloc_ids_from_const(val: ConstValue<'_>, mut f: impl FnMut(AllocId)) {
+        match val {
+            ConstValue::Scalar(interpret::Scalar::Ptr(ptr)) => {
+                f(ptr.alloc_id);
+            }
+            ConstValue::Scalar(interpret::Scalar::Raw { .. }) => {}
+            ConstValue::ByRef { alloc, .. } | ConstValue::Slice { data: alloc, .. } => {
+                for id in alloc_ids_from_alloc(alloc) {
+                    f(id);
+                }
+            }
+        }
+    }
+    struct CollectAllocIds(FxHashSet<AllocId>);
+    impl Visitor<'_> for CollectAllocIds {
+        fn visit_constant(&mut self, constant: &Constant<'tcx>, location: Location) {
+            self.super_constant(constant, location);
+            let val = match constant.literal.val {
+                ty::ConstKind::Value(val) => val,
+                _ => return,
+            };
+            alloc_ids_from_const(val, |id| {
+                self.0.insert(id);
+            });
```

Comment:
I think you want `visit_const` not `visit_constant` here and maybe even use a `TypeVisitor`.
I say that because it's possible to have an unevaluated constant with const args in its substs.